include/drivers/uart.h

```diff
@@ -467,511 +488,223 @@ __subsystem struct uart_driver_api {
 	const struct uart_driver_api *api =
-			(const struct uart_driver_api *)dev->api;
+		(const struct uart_driver_api *)dev->api;
 
-	if (api->callback_set == NULL) {
+	if (api->err_check == NULL) {
 		return -ENOSYS;
 	}
 
-	return api->callback_set(dev, callback, user_data);
-#else
-	return -ENOTSUP;
-#endif
+	return api->err_check(dev);
 }
 
 /**
- * @brief Send given number of bytes from buffer through UART.
- *
- * Function returns immediately and event handler,
- * set using @ref uart_callback_set, is called after transfer is finished.
+ * @defgroup uart_polling Polling UART API
+ * @{
+ */
+
+/**
+ * @brief Poll the device for input.
  *
- * @param dev     UART device structure.
- * @param buf     Pointer to transmit buffer.
- * @param len     Length of transmit buffer.
- * @param timeout Timeout in microseconds. Valid only if flow control is
- *		  enabled. @ref SYS_FOREVER_US disables timeout.
+ * @param dev UART device structure.
+ * @param p_char Pointer to character.
  *
- * @retval -ENOTSUP If not supported.
- * @retval -EBUSY   There is already an ongoing transfer.
- * @retval 0	    If successful, negative errno code otherwise.
+ * @retval 0 If a character arrived.
+ * @retval -1 If no character was available to read (i.e., the UART
+ *            input buffer was empty).
+ * @retval -ENOSYS If the operation is not supported.
+ * @retval -EBUSY If reception was enabled using uart_rx_enabled
  */
-__syscall int uart_tx(const struct device *dev, const uint8_t *buf,
-		      size_t len,
-		      int32_t timeout);
-
-static inline int z_impl_uart_tx(const struct device *dev, const uint8_t *buf,
-				 size_t len, int32_t timeout)
+__syscall int uart_poll_in(const struct device *dev, unsigned char *p_char);
 
+static inline int z_impl_uart_poll_in(const struct device *dev,
+				      unsigned char *p_char)
 {
-#ifdef CONFIG_UART_ASYNC_API
 	const struct uart_driver_api *api =
-			(const struct uart_driver_api *)dev->api;
+		(const struct uart_driver_api *)dev->api;
 
-	return api->tx(dev, buf, len, timeout);
-#else
-	return -ENOTSUP;
-#endif
+	if (api->poll_in == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->poll_in(dev, p_char);
 }
 
 /**
- * @brief Send given number of datum from buffer through UART.
+ * @brief Poll the device for wide data input.
  *
- * Function returns immediately and event handler,
- * set using @ref uart_callback_set, is called after transfer is finished.
+ * @param dev UART device structure.
+ * @param p_u16 Pointer to 16-bit data.
  *
- * @param dev     UART device structure.
- * @param buf     Pointer to wide data transmit buffer.
- * @param len     Length of wide data transmit buffer.
- * @param timeout Timeout in milliseconds. Valid only if flow control is
- *		  enabled. @ref SYS_FOREVER_MS disables timeout.
+ * @retval 0  If data arrived.
+ * @retval -1 If no data was available to read (i.e., the UART
+ *            input buffer was empty).
  *
  * @retval -ENOTSUP If API is not enabled.
```

Comment:
out of the scope, but this should also be `-ENOSYS`

Reply:
I think `-ENOTSUP` is correct here, as it is returned if this API is disabled by Kconfig, but it is actually implemented.

Reply:
understood